An interface that allows pre-filter plugins to update their pre-calculated status

ahg-g opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This is needed to allow efficient preemption simulations: during preemption, we remove/add pods from each node before running the filter plugins again to evaluate whether removing/ad
ding specific pods will allow the incoming pod to be scheduled on the node. Instead of calling prefilter again, we should allow the plugin to do incremental update to its pre-computed state.


**Which issue(s) this PR fixes**:
Fixes #82897

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
New APIs to allow adding/removing pods from pre-calculated prefilter state in the scheduling framework
```

draveness reviewed (commented):
/label api-review
/hold feel free to cancel

Did we have a thorough discussion of whether or not exposing these API? Is this a required interface (AddPod, RemovePod) which we must provide to the developer, or just an approach to help with the performance.

Besides, it requires an API review if we change the interface of the scheduler extension point.

ahg-g commented:
> /label api-review
> /hold feel free to cancel
> 
> Did we have a thorough discussion of whether or not exposing these API? Is this a required interface (AddPod, RemovePod) which we must provide to the developer, or just an approach to help with the performance.
Certainly, I created the issue to discuss the motivation. Will keep the hold until the community is comfortable with this addition.

> 
> Besides, it requires an API review if we change the interface of the scheduler extension point.
It does not actually.

ahg-g commented:
/cc @alculquicondor

ahg-g commented:
/test pull-kubernetes-bazel-build

Huang-Wei commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -34,22 +34,23 @@ import (
+	pluginNameToWeightMap      map[string]int
+	queueSortPlugins           []QueueSortPlugin
+	preFilterPlugins           []PreFilterPlugin
+	preFilterWithUpdatePlugins []PreFilterWithUpdatePlugin
```
I'm not sure I like the idea of separating the two plugins. In a user's perspective, what they know is just a phase called "PreFilter" which "PreFilterPlugins" should be responsible for. IMO it's better to keep update related logic into another interface.

I'd propose to keep `preFilterPlugins` only, and then add a `MetaUpdater()` method in PreFilterPlugin like this:

```go
type PreFilterPlugin interface {
	Plugin
	PreFilter(pc *PluginContext, p *v1.Pod) *Status
	MetaUpdater() MetaUpdater
}
```

And `MetaUpdater` is an interface with AddPod() and RemovePod():

```go
type MetaUpdater interface {
	AddPod(pc *PluginContext, podToSchedule *v1.Pod, podToAdd *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *Status
	RemovePod(pc *PluginContext, podToSchedule *v1.Pod, podToRemove *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *Status
}
```

So `RunPreFilter[Add|Remove]Pod` method can be updated:

```go
func (f *framework) RunPreFilterAddPod(pc *PluginContext, podToSchedule *v1.Pod,
	podToAdd *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *Status {
	for _, pl := range f.preFilterPlugins {
		status := pl.MetaUpdater().AddPod(pc, podToSchedule, podToAdd, nodeInfo)
```

alculquicondor replied:
The idea of the separate interfaces is that the extra methods are optional. And then we only iterate once to identify the plugins that implement the interface.
These private slices are not user (developer) facing.
And we have done the same for `scorePlugins` and `scoreWithNormalizePlugins`

ahg-g replied:
Right, the idea is to make the methods optional, those slices are implementation details only.

Huang-Wei replied:
Sry to confuse you. I should have commented it on the interface definitions.

So basically, instead of defining two interfaces (`PreFilterPlugin` and `PreFilterWithUpdatePlugin`), I'm leaning towards to keep one interface `PreFilterPlugin`.

With the above extra interface funtion `MetaUpdater()`, implementations that returns nil will get simply bypassed without invoking its `AddPod()` and `RemovePod()` upon preemption. Right now, we're relying on typecasting to tell whether it's a `PreFilterPlugin` or `PreFilterWithUpdatePlugin` implementation, which I'm not sure I like the way.

Huang-Wei replied:
The above suggestion is similar like what we did in GVK:

https://github.com/kubernetes/kubernetes/blob/d62ebe31f0c9bb73217acba847cc4f5f72138823/staging/src/k8s.io/apimachinery/pkg/runtime/schema/interfaces.go#L28

Basically an API object can choose not to omit GVK by implementing `GroupVersionKind()` to simply return nil, so that API machinery won't call further GVK related methods upon this object.

ahg-g replied:
this sounds good to me, it will also solve the confusion when a plugin implements only AddPod but not RemovePod for example, which will result in a failure in the type cast, and so AddPod will not be called contrary to what the developer may think.

alculquicondor replied:
Would it be worth making the optional normalize for score plugins consistent with this suggestion?

Huang-Wei replied:
I think so, but we can have it in another PR.

ahg-g replied:
Done. I named the interface "Updater" only, I think it is sufficient.

draveness replied:
> With the above extra interface funtion `MetaUpdater()`, implementations that returns nil will get simply bypassed without invoking its `AddPod()` and `RemovePod()` upon preemption. Right now, we're relying on typecasting to tell whether it's a `PreFilterPlugin` or `PreFilterWithUpdatePlugin` implementation, which I'm not sure I like the way.

I feel the same way, it's better to implement it as a noop. I sent https://github.com/kubernetes/kubernetes/pull/83042 to remove the scoreWithNormalize interface.

Huang-Wei commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -110,6 +111,14 @@ func NewFramework(r Registry, plugins *config.Plugins, args []config.PluginConfi
+
+				// Next, if the plugin also implements PodAdd/PodRemove interfaces,
+				// add it to the prefilterWithupdateplugin list.
+				np, ok := pg.(PreFilterWithUpdatePlugin)
```
If we keep `PreFilterPlugin` only, we don't need to check this.

Huang-Wei commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -306,6 +315,42 @@ func (f *framework) RunPreFilterPlugins(
+// RunPreFilterAddPod calls the AddPod interface for the set of configured
+// PreFilter plugins. It returns directly if any of the plugins return any
+// status other than Success.
+func (f *framework) RunPreFilterAddPod(pc *PluginContext, podToSchedule *v1.Pod,
```
`RunMetaAddPod` sounds more clear to me so that it's possible in the future to enable other plugins to be called under the hood.

ahg-g replied:
I don't think we make this generic, it should be specific to prefilter (i.e., the updater in the prefilter plugins). This is internal implementation in any case.

Huang-Wei replied:
SG.

Huang-Wei reviewed (commented):
Some comments below.

Huang-Wei commented:
/retest

Huang-Wei commented:
/lgtm

Thanks @ahg-g !

Huang-Wei commented:
/retest

ahg-g commented:
> /lgtm
> 
> Thanks @ahg-g !

Thanks, now that the context PR is in, I added logic to clone the context during preemption, PTAL at the last commit, if all looks good I will squash.

alculquicondor commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -1023,7 +1023,9 @@ func (g *genericScheduler) selectNodesForPreemption(
 			metaCopy = meta.ShallowCopy()
 		}
-		pods, numPDBViolations, fits := g.selectVictimsOnNode(pluginContext, pod, metaCopy, nodeNameToInfo[nodeName], fitPredicates, queue, pdbs)
+		pluginContextClone := pluginContext.Clone()
```
why not do the copy in `selectVictimsNode`, given that pluginContextClone isn't used outside of it?

Also, prefer shorter names for local variables. We could start using `pCtx` instead of `pluginContext`

ahg-g replied:
> why not do the copy in `selectVictimsNode`, given that pluginContextClone isn't used outside of it?
> 

To be consistent with how "meta" is treated.

> Also, prefer shorter names for local variables. We could start using `pCtx` instead of `pluginContext`

we already use pluginContext everywhere.

alculquicondor replied:
sg, they are just nits.

Huang-Wei commented on pkg/scheduler/framework/v1alpha1/context_test.go:
```diff
@@ -64,3 +64,11 @@ func TestPluginContextClone(t *testing.T) {
 	}
 }
+
+func TestPluginContextCloneNil(t *testing.T) {
```
Ideally, this can be merged into `TestPluginContextClone` in a Golang [subtest](https://blog.golang.org/subtests) manner.

ahg-g replied:
I feel the boilerplate for doing will outweigh its benefit.

Huang-Wei replied:
Sure, it's just nits.

Huang-Wei commented:
> Thanks, now that the context PR is in, I added logic to clone the context during preemption, PTAL at the last commit, if all looks good I will squash.

Please go squashing the commits.

ahg-g committed:
An interface that allows pre-filter plugins to update their pre-calculated.

This is needed to allow efficient preemption simulations: during preemption, we remove/add pods from each node before running the filter plugins again to evaluate whether removing/adding specific pods will allow the incoming pod to be scheduled on the node. Instead of calling prefilter again, we should allow the plugin to do incremental update to its pre-computed state.

ahg-g commented:
> > Thanks, now that the context PR is in, I added logic to clone the context during preemption, PTAL at the last commit, if all looks good I will squash.
> 
> Please go squashing the commits.

Done.

Huang-Wei commented:
/lgtm

ahg-g commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

Huang-Wei commented:
/hold cancel

tedyu commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -306,6 +306,46 @@ func (f *framework) RunPreFilterPlugins(
+	podToRemove *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *Status {
+	for _, pl := range f.preFilterPlugins {
+		if updater := pl.Updater(); updater != nil {
+			status := updater.RemovePod(pc, podToSchedule, podToRemove, nodeInfo)
```
Since this has same signature as that of updater.AddPod , it seems this func (RunPreFilterUpdaterRemovePod) can be unified with RunPreFilterUpdaterAddPod by passing AddPod / RemovePod as parameter.

ahg-g commented:
/priority important-soon
